File: apps/worker/src/email/brevo.ts
/** Email delivery through Brevo's transactional API (the pipeline uses the same one). */

const BREVO_URL = "https://api.brevo.com/v3/smtp/email";

export interface Email {
  to: string;
  subject: string;
  html: string;
  text: string;
}

export type SendEmail = (email: Email) => Promise<void>;

/** Brevo's error code only: its message may quote the recipient. */
async function errorCode(resp: Response): Promise<string> {
  const body: unknown = await resp.json().catch(() => null);
  const code = body && typeof body === "object" && "code" in body ? body.code : "";
  return typeof code === "string" ? code : "";
}

/** The transport configured by the environment: Brevo, or the logs in local development. */
export function createMailer(env: Env): SendEmail {
  if (env.EMAIL_MODE === "log") {
    return async (email) => {
      console.log(`[email] to ${email.to}: ${email.subject}\n${email.text}`);
    };
  }
  const sender = { email: env.SENDER_EMAIL, name: env.SENDER_NAME };
  const replyTo = env.REPLY_TO_EMAIL ? { replyTo: { email: env.REPLY_TO_EMAIL } } : {};
  return async (email) => {
    const resp = await fetch(BREVO_URL, {
      method: "POST",
      headers: { "api-key": env.BREVO_API_KEY, "content-type": "application/json" },
      body: JSON.stringify({
        sender,
        // No per-recipient open/click tracking: Brevo only counts them in aggregate.
        to: [{ email: email.to, contactPixelTrackingConsent: false }],
        subject: email.subject,
        htmlContent: email.html,
        textContent: email.text,
        ...replyTo,
      }),
    });
    if (!resp.ok) throw new Error(`Brevo responded ${resp.status} ${await errorCode(resp)}`);
  };
}

File: apps/pipeline/src/postthedoc/mail/brevo.py
"""Delivery through Brevo's transactional email API (the same one the Worker uses)."""

import httpx

from postthedoc.config import BrevoSettings
from postthedoc.mail.message import Email, MailError, MailerUnavailableError

BREVO_URL = "https://api.brevo.com/v3/smtp/email"
# Answers that every other email would get too: bad API key, no credits left, rate limited.
UNAVAILABLE_STATUSES = frozenset({401, 402, 403, 429})


def _error_code(resp: httpx.Response) -> str:
    """Brevo's error code, e.g. "unauthorized": its message may quote the recipient, and the
    daily job's logs are public."""
    try:
        body = resp.json()
    except ValueError:
        return ""
    code = body.get("code") if isinstance(body, dict) else None
    return code if isinstance(code, str) else ""


class BrevoMailer:
    def __init__(self, client: httpx.Client, settings: BrevoSettings) -> None:
        self._client = client
        self._api_key = settings.api_key
        self._sender = {"email": settings.sender_email, "name": settings.sender_name}
        self._reply_to = {"email": settings.reply_to} if settings.reply_to else None

    def send(self, email: Email) -> None:
        payload = {
            "sender": self._sender,
            # No per-recipient open/click tracking: Brevo only counts them in aggregate.
            "to": [{"email": email.to, "contactPixelTrackingConsent": False}],
            "subject": email.subject,
            "htmlContent": email.html,
            "textContent": email.text,
            "headers": email.headers,
        }
        if self._reply_to:
            payload["replyTo"] = self._reply_to
        try:
            resp = self._client.post(BREVO_URL, headers={"api-key": self._api_key}, json=payload)
        except httpx.HTTPError as exc:
            raise MailError(f"Brevo: {exc!r}") from exc
        if resp.is_success:
            return
        error = MailerUnavailableError if resp.status_code in UNAVAILABLE_STATUSES else MailError
        raise error(f"Brevo: HTTP {resp.status_code} {_error_code(resp)}".rstrip())
